pandas/core/reshape/merge.py

```diff
@@ -1708,6 +1708,23 @@ def _validate_specification(self):
         if self.left_by is not None and self.right_by is None:
             raise MergeError("missing right_by")
 
+        # GH#29130 Check that merge keys do not have dtype object
+        lo_dtype = (
+            self.left[self.left_on[0]].dtype
+            if not self.left_index
+            else self.left.index.dtype
+        )
+        ro_dtype = (
+            self.right[self.right_on[0]].dtype
+            if not self.right_index
+            else self.right.index.dtype
+        )
+        if is_object_dtype(lo_dtype) or is_object_dtype(ro_dtype):
+            raise ValueError(
```

Comment:
can you make this a MergeError (as everthing else is)

Reply:
Done, issue mentioned ValueError, but MergeError is probably better